File: scenes/smoke.py
"""Smoke test: drop into bash, type `tint` to launch picker, scroll, dismiss.

Run from project root:  python -m scenes.smoke
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from recorder.driver import Recorder


def main():
    r = Recorder(cols=100, rows=30)
    r.start()

    r.dwell(800, settle_ms=600)              # let bash prompt appear
    r.type_text("tint", per_char_ms=80)      # launch picker
    r.key("enter", dwell_ms=400)
    r.dwell(800)                             # picker renders

    r.key("down", dwell_ms=120, repeat=5)
    r.dwell(500)

    r.key("up", dwell_ms=120, repeat=2)
    r.dwell(500)

    r.key("escape", dwell_ms=400)
    r.dwell(600)

    r.stop()

    out = r.write_cast("assets/smoke.cast")
    print(f"wrote {out} ({out.stat().st_size} bytes, {r.event_count()} events)")


if __name__ == "__main__":
    main()

File: scenes/smoke.rs
//! Smoke scene: open picker, scroll a few rows, dismiss with Esc.
use std::path::PathBuf;
use std::time::Duration;

use clap::Parser;
use tint_recorder::recorder::{Key, Recorder, RecorderConfig};
use tint_recorder::scenes::ms;

#[derive(Parser)]
struct Args {
    #[arg(long, default_value = "assets/smoke.cast")]
    cast: PathBuf,
}

fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let mut r = Recorder::start(RecorderConfig {
        cols: 100, rows: 30, ..Default::default()
    })?;

    r.dwell(ms(800), ms(600))?;
    r.type_text("tint", ms(80))?;
    r.key(Key::Enter, ms(400))?;
    r.dwell(ms(800), ms(100))?;

    r.keys(Key::Down, ms(120), 5)?;
    r.dwell(ms(500), ms(100))?;

    r.keys(Key::Up, ms(120), 2)?;
    r.dwell(ms(500), ms(100))?;

    r.key(Key::Escape, ms(400))?;
    r.dwell(ms(600), ms(100))?;

    let cast = r.stop()?;
    let _ = std::fs::create_dir_all(args.cast.parent().unwrap_or(&PathBuf::from(".")));
    cast.write(&args.cast)?;
    println!("wrote {} ({} bytes, {} events)",
        args.cast.display(),
        std::fs::metadata(&args.cast)?.len(),
        cast.events.len(),
    );
    Ok(())
}

// Quiet a clippy lint for the unused Duration import path
const _: Option<Duration> = None;
